File: mobile/welcome-screen/ios-app/ShowUpWelcome/InputField.swift
//  InputField.swift
//  ShowUp · the one text input, and the chrome every field-shaped control wears
//
//  WHY THE WHOLE BOX BEING TAPPABLE IS THE POINT OF THIS FILE
//
//  Android shipped a number field that measured 23dp inside a 56dp row. The row was 56, the border
//  was 56, the fill was 56; the text field inside it measured to the height of its own text and the
//  row centred it. So the control looked 56 tall and only its middle third took a tap. It was
//  invisible in the source and invisible in a screenshot, and measurement found it on 15 of 18
//  phone sizes.
//
//  iOS turned out NOT to have it — `TapTargetTests` measured the field at a full 56pt at every
//  size on 8 September 2026, because a `UIViewRepresentable` with no `sizeThatFits` accepts the
//  height SwiftUI proposes. I had predicted the opposite. The guarantee is asserted rather than
//  assumed on both platforms now, because it is one modifier away from being false on either.

import SwiftUI

// MARK: - The chrome

/// One control tall, `Radius.control` corners, an elevated fill and a 1.5pt outline.
///
/// Separate from `InputField` because two controls wear it and only one of them is an input: the
/// country pill beside the number field is a Button with identical chrome. The alternative to
/// sharing it is what was there before — the same four modifiers written twice, a few lines apart,
/// already differing in their horizontal padding.
///
/// **1.5pt and `liqSubtle`, not `liqBorder`.** The outline is the only thing identifying the field,
/// and Border at 12% measures 1.28:1 against a 3:1 requirement. Audit finding 7.
struct FieldChrome: ViewModifier {
    var outline: Color = .liqSubtle
    var height: CGFloat = ComponentSizes.controlHeight
    var horizontalPadding: CGFloat = 18
    /// The soft 4pt danger ring outside the outline.
    ///
    /// **iOS only, and flagged rather than copied.** Android's invalid field has the red outline and
    /// the glyph and no ring. Adding it to Android or dropping it here both change a screen that is
    /// in PO Acceptance, so both platforms keep what they render today and the design side decides
    /// which is right. Recorded so it is a known divergence and not silent drift.
    var dangerRing: Bool = false

    func body(content: Content) -> some View {
        content
            .padding(.horizontal, horizontalPadding)
            .frame(maxWidth: .infinity, minHeight: height, maxHeight: height, alignment: .leading)
            .background(RoundedRectangle(cornerRadius: Radius.control).fill(Color.liqElevated))
            .overlay(RoundedRectangle(cornerRadius: Radius.control)
                .strokeBorder(outline, lineWidth: 1.5))
            .overlay(dangerRing ? RoundedRectangle(cornerRadius: Radius.control)
                .strokeBorder(Color.liqDanger.opacity(0.10), lineWidth: 4)
                .padding(-2.75) : nil)
    }
}

extension View {
    /// See `FieldChrome`.
    ///
    /// Takes the modifier rather than restating its parameters. The first version of this spelled
    /// out every default a second time -- `height: CGFloat = ComponentSizes.controlHeight`,
    /// `horizontalPadding: CGFloat = 18` -- and an injection test caught what that costs: changing
    /// the struct's height left the extension's copy still satisfying the check that guards it.
    /// One definition per value, which is the rule this whole task exists to apply.
    func fieldChrome(_ chrome: FieldChrome = FieldChrome()) -> some View {
        modifier(chrome)
    }
}

// MARK: - The input

/// The one text input: a single line inside `fieldChrome`, with an optional trailing adornment.
///
/// The field itself is passed in rather than built here, because the number field is a
/// `UIViewRepresentable` for a reason that has nothing to do with chrome — as-you-type grouping
/// needs caret control that SwiftUI's `TextField` does not expose. See `PhoneNumberField.swift`.
/// Wrapping it in a generic slot keeps that argument where it belongs and lets a future plain
/// `TextField` use the same box.
///
/// `label` is not optional. An input with no accessibility label is a control VoiceOver cannot name,
/// and this one had none on Android until 8 September 2026.
struct InputField<Field: View, Trailing: View>: View {
    let label: String
    var invalid: Bool = false
    @ViewBuilder var field: () -> Field
    @ViewBuilder var trailing: () -> Trailing

    var body: some View {
        HStack(spacing: 0) {
            field()
                // The counterpart of Android's fillMaxHeight. Not a parameter, so it cannot be
                // omitted -- see the note at the top of this file.
                .frame(maxHeight: .infinity)
            trailing()
        }
        .fieldChrome(FieldChrome(outline: invalid ? .liqDanger : .liqSubtle,
                                 dangerRing: invalid))
        .accessibilityElement(children: .contain)
        .accessibilityLabel(Text(label))
    }
}

extension InputField where Trailing == EmptyView {
    init(label: String, invalid: Bool = false, @ViewBuilder field: @escaping () -> Field) {
        self.init(label: label, invalid: invalid, field: field, trailing: { EmptyView() })
    }
}

/// The error glyph the number field shows when validation rejects the number.
///
/// Its own view because both platforms draw the same 22pt circle with a Lora "!" in it, and it was
/// written inline on each.
///
/// `Spacer(minLength: 0)` is copied verbatim from what this screen already renders, and not
/// "improved" to a fixed 8 to match Android's spacer. The two expressions distribute the row's
/// slack differently, this screen is in PO Acceptance, and I have already been wrong once today
/// about how SwiftUI resolves a flexible row. Preserving the expression preserves the pixels.
struct FieldErrorGlyph: View {
    var body: some View {
        Spacer(minLength: 0)
        ZStack {
            Circle().fill(Color.liqDanger).frame(width: 22, height: 22)
            Text("!").font(.custom(PS.loraBold, size: 14)).foregroundColor(.white)
        }
    }
}

File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/designsystem/InputField.kt
package com.showup.designsystem

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * The chrome every field-shaped control wears: one control tall, [Radius.control] corners, an
 * elevated fill and a 1.5dp outline.
 *
 * Exposed separately from [InputField] because two controls wear it and only one of them is an
 * input. The country pill beside the number field is a BUTTON with identical chrome, and the
 * alternative to sharing it is what was there before: the same five modifiers written twice, a few
 * lines apart, already differing in their horizontal padding.
 *
 * [outline] is a parameter rather than a boolean because the states are not symmetrical -- the
 * pill has one outline, the input has two, and a `error: Boolean` here would have to grow a case
 * every time a control gains a state.
 *
 * **1.5dp and [Subtle], not [Border].** The outline is the only thing identifying the field, and
 * Border at 12% measures 1.28:1 against a 3:1 requirement. Audit finding 7.
 */
fun Modifier.fieldChrome(
    outline: Color = Subtle,
    height: Dp = ComponentSizes.controlHeight,
    horizontalPadding: Dp = 18.dp,
    /**
     * Taken here rather than chained on by the caller, because modifier ORDER decides the tap area
     * and only one order is right.
     *
     * `.clickable().padding(14)` makes the padding part of the control -- the whole 56dp box
     * responds. `.padding(14).clickable()` looks identical and responds only inside the padding,
     * losing 28dp of width. A caller writing `.fieldChrome(...).clickable { }` would get the
     * second one, silently, which is the same class of defect as the 23dp field this file exists
     * to prevent. So the hook is a parameter and the order is not the caller's to get wrong.
     */
    onClick: (() -> Unit)? = null,
): Modifier {
    val shape = RoundedCornerShape(Radius.control)
    return this
        .height(height)
        .clip(shape)
        .background(Elevated)
        .border(1.5.dp, outline, shape)
        .then(
            if (onClick != null) Modifier.clickable(role = Role.Button, onClick = onClick)
            else Modifier
        )
        .padding(horizontal = horizontalPadding)
}

/**
 * The one text input. A single line inside [fieldChrome], with a placeholder, an optional trailing
 * adornment, and the whole box tappable.
 *
 * WHY THE WHOLE BOX BEING TAPPABLE IS THE POINT OF THIS FILE
 *
 * This project shipped a number field that measured **23dp inside a 56dp row**. The row was 56, the
 * border was 56, the fill was 56; the text field inside it measured to the height of its own text
 * and the row centred it. So the control looked 56dp tall and only its middle third would take a
 * tap -- touching near the top or bottom edge did nothing at all. It was invisible in the source
 * and invisible in a screenshot, and ScreenFitTest found it on 15 of 18 phone sizes.
 *
 * The fix was one modifier. That is exactly why it belongs here instead of at a call site: a
 * `fillMaxHeight()` in a screen is one careless edit from being dropped, and the next person to add
 * an input starts from nothing and can make the same mistake from scratch. Here there is no way to
 * ask for a field that does not fill its box -- the modifier is not a parameter -- and
 * `TapTargetTest` asserts the measured height on both platforms.
 *
 * Because the field fills the box, its text has to be centred by [decorationBox] rather than
 * riding the top edge.
 */
@Composable
fun InputField(
    value: String,
    onValueChange: (String) -> Unit,
    /** Spoken by TalkBack. Required: an unlabelled input is a control a screen reader cannot name. */
    label: String,
    placeholder: String,
    modifier: Modifier = Modifier,
    /** Red outline plus whatever the caller puts in [trailing]. */
    invalid: Boolean = false,
    keyboardType: KeyboardType = KeyboardType.Phone,
    /**
     * What the field holds, so the keyboard can offer to fill it. Null means "do not offer".
     *
     * Explicit rather than derived from [keyboardType], because the two answer different questions:
     * a phone keypad is also what a PIN and a verification code are typed on, and inferring
     * [FieldContent.PhoneNumber] from [KeyboardType.Phone] would offer somebody's phone number as a
     * PIN.
     */
    contentType: FieldContent? = null,
    onSubmit: () -> Unit = {},
    focusRequester: FocusRequester? = null,
    visualTransformation: VisualTransformation = VisualTransformation.None,
    /** The error glyph on the number field. Sits inside the box, after the text. */
    trailing: (@Composable () -> Unit)? = null,
) {
    val style = TextStyle(
        color = Fg, fontFamily = Manrope, fontWeight = FontWeight.SemiBold,
        fontSize = 17.sp, letterSpacing = 0.3.sp,
    )
    Row(
        modifier.fieldChrome(outline = if (invalid) Danger else Subtle),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            // fillMaxHeight is the whole point -- see the note above. Not a parameter, so it
            // cannot be omitted.
            modifier = Modifier
                .weight(1f)
                .fillMaxHeight()
                // Before the focus requester, so the observer is above the target it watches. A
                // filled value goes through the caller's onValueChange rather than around it,
                // which is what keeps the number field's digit filter and length cap applying to
                // an autofilled number exactly as they do to a typed one.
                .then(
                    contentType?.let { type ->
                        Modifier.autofill(type) { onValueChange(it) }
                    } ?: Modifier
                )
                .then(focusRequester?.let { Modifier.focusRequester(it) } ?: Modifier)
                .semantics { contentDescription = label },
            textStyle = style,
            singleLine = true,
            cursorBrush = SolidColor(Purple),
            keyboardOptions = KeyboardOptions(
                keyboardType = keyboardType, imeAction = ImeAction.Done,
            ),
            // Enter submits, so the keyboard never has to be dismissed to reach the CTA.
            keyboardActions = KeyboardActions(onDone = { onSubmit() }),
            visualTransformation = visualTransformation,
            decorationBox = { inner ->
                Box(contentAlignment = Alignment.CenterStart) {
                    if (value.isEmpty()) {
                        Text(
                            placeholder, color = Faint, fontFamily = Manrope,
                            fontWeight = FontWeight.SemiBold, fontSize = 17.sp,
                            letterSpacing = 0.3.sp, maxLines = 1,
                        )
                    }
                    inner()
                }
            },
        )
        if (trailing != null) {
            // An explicit spacer, not spacedBy: with one child spacedBy puts no space anywhere,
            // and the gap between the digits and the error glyph is 8 in the design.
            Spacer(Modifier.width(Spacing.md))
            trailing()
        }
    }
}
